### PR TITLE
vsr: state sync performance improvements

### PR DESCRIPTION
**Pull request by chaitanyabhandari**
### Summary

This PR improves the performance of state sync (specifically, the syncing of table blocks) using non-intrusive optimizations and configuration changes. Using some simple observations, we were able to improve the time that it takes to sync a ~86G data file from **_13095 seconds (~218 minutes) → 869s (~14.5 minutes)_**. 

These numbers are from running the following simple experiment on a _3-node TB cluster on t2.large AWS instances_:

1. Start R1, R2, but not R0
2. Run tigerbeetle benchmark with `transfer_count=150,000,000`, wait for it to finish
3. Start R0, measure time taken to sync all the content with R1 & R2's help

### Optimizations

Here are the optimizations that we implemented:

1. https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2286193309 explains why we removed this optimization: ~**Commit pipeline (and thereby compaction) paused till content is synced**: The motivation behind pausing the commit pipeline (and thereby compaction) is that we observed a lot of `error(grid): 0: invalid_checksum: expected address=X checksum=Y, found address=0 checksum=0` while table blocks were being synced. This is because compaction tries to read index/data blocks, finds them empty, and queues the block(s) for repair. During state sync, reading a block (and incurring extraneous IO) is a rather roundabout way to to discover that a block needs to synced, when in fact [sync_enqueue_tables](https://github.com/tigerbeetle/tigerbeetle/blob/918c25070e3b479fec561ff00bd1f35c0d6df2a4/src/vsr/replica.zig#L9111) is already doing the same without incurring IO.~ 
2. **Proactive request_blocks**: Inspired by  [this optimization that we have for request_prepares](https://github.com/tigerbeetle/tigerbeetle/blob/c88e34235719a45a8a3f3347f35d9d9d68baef49/src/vsr/replica.zig#L9654-L9657)), we now [proactively send request_blocks](https://github.com/tigerbeetle/tigerbeetle/commit/d14a1f986f12e721742c0734cad6f2a8aea55e01) before the next timeout, in case the syncing replica has enough write iops.
3. **Increasing grid_missing_tables_max from 3 → 6**: Increases the # of tables that are concurrently synced

<details>
<summary>Here is the performance improvement that each of the aforementioned optimizations brings:</summary>

#### No optimizations

```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "sync: ops\|sync_enqueue_tables\|lagging"
warning(replica): 0: on_start_view: start sync; lagging behind cluster (op_prepare_max=999 quorum_head=18420)
info(replica): 0: sync: ops=0..18271
info(replica): 0: sync_enqueue_tables: all tables synced (commit=0..18271) in 13095s
```

#### Compaction Freeze: 

```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "sync: ops\|sync_enqueue_tables: \|lagging"
warning(replica): 0: on_start_view: start sync; lagging behind cluster (op_prepare_max=999 quorum_head=18420)
info(replica): 0: sync: ops=0..18271
info(replica): 0: sync_enqueue_tables: all tables synced (commit=0..18271) in 11034s
```

#### Compaction Freeze + Proactive request_blocks

```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "sync: ops\|sync_enqueue_tables: \|lagging"
warning(replica): 0: on_start_view: start sync; lagging behind cluster (op_prepare_max=999 quorum_head=18420)
info(replica): 0: sync: ops=0..18271
info(replica): 0: sync_enqueue_tables: all tables synced (commit=0..18271) in 1054s
```

#### Compaction Freeze + Proactive request_blocks + grid_missing_tables_max: 3 → 6

```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "sync: ops\|sync_enqueue_tables\|lagging"
warning(replica): 0: on_start_view: start sync; lagging behind cluster (op_prepare_max=999 quorum_head=18420)
info(replica): 0: sync: ops=0..18271
info(replica): 0: sync_enqueue_tables: all tables synced (commit=0..18271) in 869s
```
</details>

### CODE REVIEW

**Comment by chaitanyabhandari** [comment]
From https://github.com/tigerbeetle/tigerbeetle/actions/runs/10362913189/job/28685691012?pr=2207, realized that a few replica tests which test state sync assert that the syncing replica is also able to commit while tables are being synced. We changed this in https://github.com/tigerbeetle/tigerbeetle/pull/2207/commits/971090d9cd790abdd52be75306b832661325717f, wherein we pause then commit pipeline before tables are synced. Before changing all replica tests to reflect this new behaviour, I want to do one more experiment, where we _don't_ pause the commit pipeline, but we do apply the other two optimizations _(Proactive request_blocks + grid_missing_tables_max: 3 → 6)_. If the effect of pausing commit pipeline is negligible in the presence of the other two optimizations, we can get away with not changing this behavior at all.

File: src/vsr/replica.zig [hunk 1]
```diff
@@ -2869,14 +2869,22 @@ pub fn ReplicaType(
...
             self.sync_reclaim_tables();
+            defer {
+                self.grid_repair_writes.release(write);
+                // Proactively send another request_blocks request if there are enough write IOPs.
+                if (self.grid.callback != .cancel and
+                    self.grid_repair_writes.available() >= constants.grid_repair_request_max)
+                {
+                    self.send_request_blocks();
+                }
+            }
```
Comment (matklad) [inline src/vsr/replica.zig:2881]: I'd either keep the refer where it was, or, if I am moving the code here, got rid of defer

**Comment by chaitanyabhandari** [comment]
Re https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2285271858: Hunch confirmed! The benefits of pausing the commit pipeline are indeed diminished by the other two optimizations. We're still able to sync in **_806s._** So, we can get continue to embrace the concurrency here! I will remove https://github.com/tigerbeetle/tigerbeetle/pull/2207/commits/971090d9cd790abdd52be75306b832661325717f.

```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "sync: ops\|sync_enqueue_tables: \|lagging"
info(replica): 0: sync: ops=0..18271
warning(replica): 0: on_repair_sync_timeout: request sync; lagging behind cluster (op_head=18320 commit_min=18239 commit_max=18320 commit_stage=idle)
info(replica): 0: sync_enqueue_tables: all tables synced (commit=0..18271) in 806s
```
cc: @matklad @jorangreef

File: src/config.zig [hunk 2]
```diff
@@ -131,10 +131,10 @@ const ConfigProcess = struct {
     grid_iops_read_max: u64 = 16,
     grid_iops_write_max: u64 = 16,
     grid_cache_size_default: u64 = 1024 * 1024 * 1024,
-    grid_repair_request_max: usize = 8,
-    grid_repair_reads_max: usize = 8,
+    grid_repair_request_max: usize = 4,
+    grid_repair_reads_max: usize = 4,
     grid_missing_blocks_max: usize = 30,
-    grid_missing_tables_max: usize = 3,
+    grid_missing_tables_max: usize = 6,
```
Comment (matklad) [inline src/config.zig:137]: Might add a comment here explaining that this particular numbers have _some_ basis in benchmarks, to make it clear which parameters actually do have a “why” behind them, and which are just “TODO: figure out optimal constant”

**Review by matklad — COMMENTED** [review]
This looks basically good to me me, but I’d leave review to dj!

**Comment by chaitanyabhandari** [comment]
To get a feel of whether we're close to line rate (the global maxima for the performance we can achieve), we compare our performance with rsync. cc: @jorangreef 

**PS: We will repeat these experiments using _iperf_, since results with SSH + rsync could be less reliable than _iperf_, due to high CPU overhead (and potentially single core usage) of SSH. Therefore, our global maxima for state sync performance could be higher. Thanks @cb22 for recommending this on [Slack](https://tigerbeetleinc.slack.com/archives/C042KLT9R8F/p1723628994095389)! 🚀**

### Machine Specifications

These experiments were conducted on [t2.large](https://aws.amazon.com/ec2/instance-types/t2/) instances with 8G RAM and 2 cores:
```C
ubuntu@ip-172-31-3-127:~/tigerbeetle$ free -mh
               total        used        free      shared  buff/cache   available
Mem:           7.8Gi       3.6Gi       167Mi       940Ki       4.3Gi       4.2Gi
Swap:             0B          0B          0B

ubuntu@ip-172-31-3-127:~/tigerbeetle$ lscpu
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          46 bits physical, 48 bits
                          virtual
  Byte Order:             Little Endian
CPU(s):                   2
  On-line CPU(s) list:    0,1
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Xeon(R) CPU E5-26
                          86 v4 @ 2.30GHz
    CPU family:           6
    Model:                79
    Thread(s) per core:   1
    Core(s) per socket:   2
```

### Result

To sync **_~52.5G_** worth of data blocks, state sync takes **_806 seconds_** (https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2286193309), while rsync takes **_628.89 seconds_**.

### Details 

In the process of syncing a ~86G data file, the syncing replica repaired a total of 107432 blocks, i.e. ~52.5G worth of blocks. This takes **_806 seconds_** (https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2286193309).
```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "on_block" | wc -l
107432
```

rsync-ing the ~56G data file from R1 → R2, where R2 doesn't have any data file takes **_640 seconds_** (628.89 seconds if we extrapolate the bandwidth rsync gets to ~52.5G):
```C
ubuntu@ip-172-31-3-127:~/tigerbeetle$ rsync -avmPz -e "ssh -i chaitanyabhandari.pem" data.tigerbeetle ubuntu@172.31.3.11:/home/ubuntu/tigerbeetle/data.tigerbeetle
building file list ...
1 file to consider
data.tigerbeetle
 56,074,723,328 100%   83.48MB/s    0:10:40 (xfr#1, to-chk=0/1)

sent 17,954,925,730 bytes  received 35 bytes  27,988,972.35 bytes/sec
total size is 56,074,723,328  speedup is 3.12
```

**Review by sentientwaffle — APPROVED** [review]
LGTM!

**Comment by chaitanyabhandari** [comment]
To get a feel of whether we're close to line rate (the global maxima for the performance we can achieve), we compare our performance with iperf:



### Result

To sync **_~52.5G_** worth of data blocks, state sync takes **_806 seconds_** (https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2286193309), while rsync takes **_701 seconds_**. 

### Machine Specifications

These experiments were conducted on [t2.large](https://aws.amazon.com/ec2/instance-types/t2/) instances with 8G RAM and 2 cores:
```C
ubuntu@ip-172-31-3-127:~/tigerbeetle$ free -mh
               total        used        free      shared  buff/cache   available
Mem:           7.8Gi       3.6Gi       167Mi       940Ki       4.3Gi       4.2Gi
Swap:             0B          0B          0B

ubuntu@ip-172-31-3-127:~/tigerbeetle$ lscpu
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          46 bits physical, 48 bits
                          virtual
  Byte Order:             Little Endian
CPU(s):                   2
  On-line CPU(s) list:    0,1
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Xeon(R) CPU E5-26
                          86 v4 @ 2.30GHz
    CPU family:           6
    Model:                79
    Thread(s) per core:   1
    Core(s) per socket:   2
```

### Details 

In the process of syncing a ~86G data file, the syncing replica repaired a total of 107432 blocks, i.e. ~52.5G worth of blocks. This takes **_806 seconds_** (https://github.com/tigerbeetle/tigerbeetle/pull/2207#issuecomment-2286193309).
```C
ubuntu@ip-172-31-3-11:~/tigerbeetle$ cat replica0.vopr | grep "on_block" | wc -l
107432
```
Sending a ~52.2G of data using iperf takes **_701 seconds_**:
```C
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-701.00 sec  52.2 GBytes   640 Mbits/sec  126             sender
        Sent 52.2 GByte / 52.2 GByte (100%) of data.tigerbeetle
[  5]   0.00-701.00 sec  52.2 GBytes   640 Mbits/sec                  receiver

iperf Done.
```

**Pull request merged**